
### 1. Mathematics for AI/ML: Module 1 — Real Numbers & Number System

#### Introduction

The Number System forms the foundational bedrock of all quantitative computing. In Machine Learning and Deep Learning, continuous values, weights, gradients, and loss functions rely heavily on the properties of real numbers. Every tensor operation in PyTorch or TensorFlow is fundamentally a high-dimensional operation over subsets of real numbers.

Understanding the classification, precision, and continuum of numbers allows engineers to reason about gradient flow, numerical stability, quantization, and floating-point representations in neural networks.

#### Motivation

Computers do not possess infinite memory or infinite precision. When training modern Deep Learning architectures (like Large Language Models), we map continuous mathematical spaces into discrete digital representations (e.g., FP32, FP16, INT8). Understanding real numbers and their computational limits prevents critical bugs like vanishing/exploding gradients, numerical overflow, underflow, and precision loss during model quantization.

#### Standard Definition

The set of **Real Numbers** ($\mathbb{R}$) is the union of rational numbers ($\mathbb{Q}$) and irrational numbers ($\mathbb{I}$). Mathematically, $\mathbb{R}$ is a complete, ordered Archimedean field that corresponds to every point on an unbroken continuous line (the real line).

#### Theory

The hierarchy of number systems is defined nestedly:

* **Natural Numbers ($\mathbb{N}$):** Counting numbers $\{1, 2, 3, \dots\}$.
* **Whole Numbers ($\mathbb{W}$):** Natural numbers including zero $\{0, 1, 2, 3, \dots\}$.
* **Integers ($\mathbb{Z}$):** Whole numbers and their negative counterparts $\{\dots, -2, -1, 0, 1, 2, \dots\}$.
* **Rational Numbers ($\mathbb{Q}$):** Numbers expressible as $\frac{a}{b}$, where $a, b \in \mathbb{Z}$ and $b \neq 0$.
* **Irrational Numbers ($\mathbb{I}$):** Numbers that cannot be expressed as a simple fraction (non-repeating, non-terminating decimals, e.g., $\pi, e, \sqrt{2}$).
* **Real Numbers ($\mathbb{R}$):** $\mathbb{R} = \mathbb{Q} \cup \mathbb{I}$.
* **Complex Numbers ($\mathbb{C}$):** Numbers of the form $a + bi$, where $a, b \in \mathbb{R}$ and $i = \sqrt{-1}$.

```
Complex Numbers (C)
  └── Real Numbers (R)
        ├── Rational Numbers (Q)
        │     ├── Integers (Z)
        │     │     └── Whole Numbers (W)
        │     │           └── Natural Numbers (N)
        │     └── Non-integer Fractions
        └── Irrational Numbers (I)

```

#### Mathematical Properties

For any $a, b, c \in \mathbb{R}$:

1. **Closure Property:** $a + b \in \mathbb{R}$ and $a \cdot b \in \mathbb{R}$.
2. **Commutative Property:** $a + b = b + a$ and $a \cdot b = b \cdot a$.
3. **Associative Property:** $(a + b) + c = a + (b + c)$ and $(a \cdot b) \cdot c = a \cdot (b \cdot c)$.
4. **Distributive Property:** $a \cdot (b + c) = a \cdot b + a \cdot c$.
5. **Identity Element:** $a + 0 = a$ (Additive Identity: 0) and $a \cdot 1 = a$ (Multiplicative Identity: 1).
6. **Density Property:** Between any two distinct real numbers, there exists infinitely many real numbers.

#### Derivations

**Proof by Contradiction: $\sqrt{2}$ is Irrational**

1. Assume $\sqrt{2}$ is rational. Then $\sqrt{2} = \frac{a}{b}$, where $a, b \in \mathbb{Z}$, $b \neq 0$, and $\gcd(a, b) = 1$ (coprime).
2. Squaring both sides: $2 = \frac{a^2}{b^2} \implies a^2 = 2b^2$.
3. Thus, $a^2$ is even, which implies $a$ must be even. Let $a = 2k$ for some integer $k$.
4. Substitute $a = 2k$: $(2k)^2 = 2b^2 \implies 4k^2 = 2b^2 \implies b^2 = 2k^2$.
5. Thus, $b^2$ is even, which implies $b$ must also be even.
6. If both $a$ and $b$ are even, they share a common factor of 2. This contradicts $\gcd(a, b) = 1$.
7. Therefore, $\sqrt{2}$ must be irrational ($\sqrt{2} \in \mathbb{I}$).

#### Visualization / Intuition

Imagine a continuous, unbroken line extending infinitely in both directions. Rational numbers populate the line densely, but leave infinitesimal "holes." Irrational numbers fill every single hole completely, making the line continuous (the **Completeness Axiom**). In deep learning, loss landscapes are treated as continuous surfaces over $\mathbb{R}^n$, allowing gradient calculus to compute smooth slopes everywhere.

#### Real-world Examples

* Measuring physical continuous properties such as temperature, speed, and time.
* Audio signals represented as continuous sound waves over time.

#### AI/ML Applications

* **Model Parameters:** Neural network weights and biases are initialized as real numbers sampled from continuous probability distributions (e.g., Gaussian distribution).
* **Precision Formats:**
* `FP32` (Single-precision float): 32-bit real number representation.
* `FP16` / `BF16` (Half-precision): Used in modern LLM training to double throughput and halve VRAM requirements.
* `INT8` Quantization: Mapping continuous $\mathbb{R}$ values to discrete 8-bit integers to optimize edge deployment.



#### Python Implementation

In [4]:
pip install numpy 

Note: you may need to restart the kernel to use updated packages.


In [5]:
import numpy as np

# Real number representation and precision limits in ML
float_32 = np.float32(1.0 / 3.0)
float_64 = np.float64(1.0 / 3.0)

print(f"FP32 Precision: {float_32:.18f}")
print(f"FP64 Precision: {float_64:.18f}")

# Simulating floating point underflow (numerical instability)
eps = np.finfo(np.float32).eps
print(f"Machine Epsilon (FP32): {eps}")

# Quantization intuition: Mapping continuous float to discrete int8
float_weights = np.array([-1.5, 0.0, 0.75, 2.3], dtype=np.float32)
scale = 127.0 / np.max(np.abs(float_weights))
int8_quantized = np.round(float_weights * scale).astype(np.int8)

print(f"Original Float Weights: {float_weights}")
print(f"Quantized INT8 Weights: {int8_quantized}")

FP32 Precision: 0.333333343267440796
FP64 Precision: 0.333333333333333315
Machine Epsilon (FP32): 1.1920928955078125e-07
Original Float Weights: [-1.5   0.    0.75  2.3 ]
Quantized INT8 Weights: [-83   0  41 127]


#### Output
#### Interview Questions

1. **Q:** Why can't computers store exact real numbers, and how does this affect gradient descent?
* **A:** Real numbers are uncountable and infinite, while computer memory is finite. Floating-point approximations introduce rounding errors. In gradient descent, extreme precision loss can lead to vanishing gradients or sub-optimal convergence.


2. **Q:** What is the key difference between FP16 and BF16 in deep learning?
* **A:** Both use 16 bits, but BF16 allocates 8 bits to the exponent (same dynamic range as FP32) and 7 to the mantissa, preventing underflow/overflow without needing dynamic loss scaling during model training.



#### Common Mistakes

* Assuming floating-point arithmetic is exact (e.g., in Python, `0.1 + 0.2 == 0.3` evaluates to `False`).
* Using standard FP32 operations without checking for `NaN` (Not a Number) or `Inf` when dividing by near-zero values.

#### Memory Tricks

* **N-W-I-R-C (Never Walk In Rain Casual):** **N**atural $\subset$ **W**hole $\subset$ **I**ntegers $\subset$ **R**eal $\subset$ **C**omplex.

#### Summary

Real numbers ($\mathbb{R}$) cover both rational and irrational numbers to form a continuous number line. In AI/ML, continuous spaces enable gradient calculus, while computational constraints require mastering floating-point precision, overflow, and quantization strategies.

#### Practice Questions

1. Prove that $\log_2(3)$ is an irrational number.
2. Given a float array $W = [-2.4, 1.2, 0.0, 3.6]$, calculate the INT8 scale factor $S = \frac{127}{\max(\vert{}W\vert{})}$ and compute the quantized integer array.

---

### How to Proceed

We can walk through this complete roadmap sequentially, topic by topic or module by module.

For a comprehensive video overview of structuring an AI engineering learning path, see this [Modern AI/ML Roadmap Video](https://www.youtube.com/watch?v=YL0xu8Z73v4). This video is relevant as it discusses layer-by-layer roadmap strategies for learning AI, ML, and MLOps.